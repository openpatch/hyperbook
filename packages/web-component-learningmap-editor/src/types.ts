import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface CompletionNeed {
  id: string;
  source?: string;
  target?: string;
}

export interface Completion {
  needs?: CompletionNeed[];
  optional?: CompletionNeed[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  background?: BackgroundConfig;
  edgeConfig?: EdgeConfig;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
