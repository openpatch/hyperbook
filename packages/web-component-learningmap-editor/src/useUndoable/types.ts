export type ActionType =
  | "undo"
  | "redo"
  | "update"
  | "reset"
  | "resetInitialState";

export type HistoryLimit = number | "infinium" | "infinity";

export type MutationBehavior =
  | "mergePastReversed"
  | "mergePast"
  | "destroyFuture"
  | "keepFuture";

export interface Action<T> {
  type: ActionType;
  payload?: T;
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
  ignoreAction?: boolean;
}

export interface State<T> {
  past: T[];
  present: T;
  future: T[];
}

export interface Options {
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
}

export type UseUndoable<T> = [
  T,
  (
    payload: T | ((oldValue: T) => T),
    behavior?: MutationBehavior,
    ignoreAction?: boolean,
  ) => void,
  {
    past: T[];
    future: T[];

    undo: () => void;
    canUndo: boolean;
    redo: () => void;
    canRedo: boolean;

    reset: (initialState?: T) => void;
    resetInitialState: (newInitialState: T) => void;
    static_setState: (
      payload: T,
      behavior?: MutationBehavior,
      ignoreAction?: boolean,
    ) => void;
  },
];
