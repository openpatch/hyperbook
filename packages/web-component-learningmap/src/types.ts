import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface Completion {
  needs?: string[];
  optional?: string[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface Settings {
  title?: string;
  background?: BackgroundConfig;
  language?: string;
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  settings: Settings;
  version: number;
}

export interface RoadmapState {
  nodes: Record<string, { state: string }>;
  x: number;
  y: number;
  zoom: number;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
