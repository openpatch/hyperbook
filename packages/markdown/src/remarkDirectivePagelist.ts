// Register directive nodes in mdast:
/// <reference types="mdast-util-directive" />
//
import {
  HyperbookContext,
  HyperbookPage,
  HyperbookSection,
} from "@hyperbook/types";
import handlebars from "handlebars";
import fs from "fs";
import { Root } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import { Element } from "hast";
import {
  expectLeafDirective,
  isDirective,
  registerDirective,
} from "./remarkHelper";
import path from "path";
import { toHast } from "mdast-util-to-hast";
import { type } from "os";
import { table } from "console";

const getPageList = (
  sections: HyperbookSection[],
  pages: HyperbookPage[],
): HyperbookPage[] => {
  let pageList = [...pages];

  for (const section of sections) {
    pageList = [
      ...pageList,
      section,
      ...getPageList(section.sections, section.pages),
    ];
  }

  return pageList;
};

const clusterPages = (
  pages: HyperbookPage[],
): Record<string, HyperbookPage[]> => {
  const cluster: Record<string, HyperbookPage[]> = {};

  for (let page of pages) {
    const letter = page.name[0].toUpperCase();
    if (!cluster[letter]) {
      cluster[letter] = [];
    }

    cluster[letter].push(page);
  }

  return cluster;
};

export default (ctx: HyperbookContext) => () => {
  const name = "pagelist";
  return (tree: Root, file: VFile) => {
    const pageList = getPageList(ctx.navigation.sections, ctx.navigation.pages);
    pageList.push(...ctx.navigation.glossary);
    visit(tree, function (node) {
      if (isDirective(node)) {
        if (node.name !== name) return;

        const data = node.data || (node.data = {});
        const attributes = node.attributes || {};
        const {
          format = "ul",
          orderBy = "name:asc",
          source = "href(.*)",
        } = attributes;

        expectLeafDirective(node, file, name);
        registerDirective(file, name, [], ["style.css"]);

        data.hName = "div";
        data.hProperties = {
          class: "directive-pagelist",
        };

        const sources = (source?.split(" AND ") || []).map((s) => ({
          type: s.split("(")[0],
          regex: s.split("(")[1].split(")")[0] || "",
        }));

        let filteredPages = pageList.filter((p) => {
          let accept = true;
          for (let source of sources) {
            if (source.type === "href") {
              accept = accept && p.href?.match(source.regex) !== null;
            } else if (source.type === "name") {
              accept = accept && p.name.match(source.regex) !== null;
            } else if (source.type === "keyword") {
              let foundOne = false;
              for (let keyword of p.keywords || []) {
                if (keyword.match(source.regex)) {
                  foundOne = true;
                  break;
                }
              }
              accept = accept && foundOne;
            }
          }
          return accept;
        });

        const orderByKey = orderBy?.split(":")[0] || "name";
        const orderByMode = orderBy?.split(":")[1] || "desc";

        filteredPages = filteredPages.sort((p1, p2) => {
          let result = 0;
          if (orderByKey === "name") {
            result = p1.name.localeCompare(p2.name);
          } else if (orderByKey === "href") {
            result = (p1.href || "").localeCompare(p2.href || "");
          } else if (orderByKey === "index") {
            result = (p1.index || 0) - (p2.index || 0);
          }

          if (orderByMode === "desc") {
            return result * -1;
          }

          return result;
        });

        filteredPages = filteredPages.map((p) => ({
          ...p,
          href: ctx.makeUrl(p.href || "", "book"),
        }));

        if (format?.startsWith("#")) {
          const snippetId = format.slice(1);
          const snippetFile = fs.readFileSync(
            path.join(ctx.root, "snippets", snippetId + ".md.hbs"),
            { encoding: "utf8" },
          );
          const template = handlebars.compile(snippetFile);
          const content = template({ pages: filteredPages });

          data.hChildren = [
            toHast(fromMarkdown(content), {
              allowDangerousHtml: true,
            }) as Element,
          ];
        } else if (format === "ul") {
          data.hChildren = [
            {
              type: "element",
              tagName: "ul",
              properties: {},
              children: filteredPages.map((p) => ({
                type: "element",
                tagName: "li",
                properties: {},
                children: [
                  {
                    type: "element",
                    tagName: "a",
                    properties: {
                      href: p.href,
                    },
                    children: [
                      {
                        type: "text",
                        value: p.name,
                      },
                    ],
                  },
                ],
              })),
            },
          ];
        } else if (format === "ol") {
          data.hChildren = [
            {
              type: "element",
              tagName: "ol",
              properties: {},
              children: filteredPages.map((p) => ({
                type: "element",
                tagName: "li",
                properties: {},
                children: [
                  {
                    type: "element",
                    tagName: "a",
                    properties: {
                      href: p.href,
                    },
                    children: [
                      {
                        type: "text",
                        value: p.name,
                      },
                    ],
                  },
                ],
              })),
            },
          ];
        } else if (format === "glossary") {
          const cluster = clusterPages(filteredPages);

          data.hChildren = Object.keys(cluster)
            .sort((a, b) => a.localeCompare(b))
            .map((letter) => ({
              type: "element",
              tagName: "div",
              properties: {
                class: "container",
              },
              children: [
                {
                  type: "element",
                  tagName: "div",
                  properties: {
                    class: "letter",
                  },
                  children: [
                    {
                      type: "text",
                      value: letter,
                    },
                  ],
                },
                {
                  type: "element",
                  tagName: "ul",
                  properties: {
                    class: "terms",
                  },
                  children: cluster[letter].map((term) => ({
                    type: "element",
                    tagName: "li",
                    properties: {},
                    children: [
                      {
                        type: "element",
                        tagName: "a",
                        properties: {
                          class: "term",
                          href: term.href,
                        },
                        children: [
                          {
                            type: "text",
                            value: term.name,
                          },
                        ],
                      },
                    ],
                  })),
                },
              ],
            }));
        }
      }
    });
  };
};
