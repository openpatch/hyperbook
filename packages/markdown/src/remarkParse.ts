// @ts-nocheck
import { Root } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { Plugin } from "unified";
import micromarkSubSup from "./micromarkSubSup";
import mdastSubSup from "./mdastSubSup";

const remarkParse: Plugin<[], string, Root> = function () {
  const self = this;

  self.parser = function (doc) {
    const extensions = self.data("micromarkExtensions") || [];
    const mdastExtensions = self.data("fromMarkdownExtensions") || [];

    return fromMarkdown(doc, {
      ...self.data("settings"),
      extensions: [micromarkSubSup, ...extensions],
      mdastExtensions: [mdastSubSup, ...mdastExtensions],
    });
  };
};

export default remarkParse;
