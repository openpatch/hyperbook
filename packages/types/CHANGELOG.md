# @hyperbook/types

## 0.0.1

### Patch Changes

- [#316](https://github.com/openpatch/hyperbook/pull/316) [`cedb551`](https://github.com/openpatch/hyperbook/commit/cedb55191fd025b5a214df406a53cbab5d1b1bc1) Thanks [@mikebarkmin](https://github.com/mikebarkmin)! - Extract Hyperbook types into a separate package
