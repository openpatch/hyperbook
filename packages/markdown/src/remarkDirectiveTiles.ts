// Register directive nodes in mdast:
/// <reference types="mdast-util-directive" />
//
import { HyperbookContext } from "@hyperbook/types";
import { Comment, Element, Text } from "hast";
import { Root } from "mdast";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import {
  expectContainerDirective,
  expectLeafDirective,
  isDirective,
  registerDirective,
} from "./remarkHelper";
import { Raw } from "mdast-util-to-hast";

export default (ctx: HyperbookContext) => () => {
  return (tree: Root, file: VFile) => {
    visit(tree, function (node) {
      if (isDirective(node)) {
        if (node.name !== "tiles") return;

        const data = node.data || (node.data = {});

        if (node.name === "tiles") {
          expectContainerDirective(node, file, "tiles");
          data.hName = "div";
          data.hProperties = {
            class: "directive-tiles",
          };
          registerDirective(file, "tiles", [], ["style.css"], ["tile"]);

          const tilesChildren: (Text | Comment | Element | Raw)[] = [];
          node.children.filter(isDirective).forEach((node) => {
            if (node.name !== "tile") {
              return;
            }
            const { title, size = "M", icon, href } = node.attributes || {};
            expectLeafDirective(node, file, "tile");

            if (!title) {
              file.fail(
                `[tiles] A title is required (${ctx.navigation.current?.href})`,
                node,
              );
            }

            const tileChildren: (Text | Comment | Element | Raw)[] = [
              {
                type: "element",
                tagName: "a",
                properties: {
                  class: "tile-title",
                  href: href ? ctx.makeUrl(href, "public") : undefined,
                },
                children: [
                  {
                    type: "text",
                    value: title,
                  },
                ],
              },
            ];

            if (icon) {
              tileChildren.push({
                type: "element",
                tagName: "img",
                properties: {
                  class: "tile-icon",
                  src: ctx.makeUrl(icon, "public"),
                },
                children: [],
              });
            }

            tilesChildren.push({
              type: "element",
              tagName: "li",
              properties: {
                class: `tile ${size} ${href ? "link" : ""}`,
              },
              children: tileChildren,
            });
          });

          data.hChildren = [
            {
              type: "element",
              tagName: "ul",
              properties: {
                class: "tiles",
              },
              children: tilesChildren,
            },
          ];
        }
      }
    });
  };
};
