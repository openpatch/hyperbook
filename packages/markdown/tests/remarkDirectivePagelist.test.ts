import { HyperbookContext } from "@hyperbook/types/dist";
import { describe, expect, it } from "vitest";
import rehypeStringify from "rehype-stringify";
import remarkParse from "remark-parse";
import remarkToRehype from "remark-rehype";
import rehypeFormat from "rehype-format";
import { unified, PluggableList } from "unified";
import remarkDirective from "remark-directive";
import remarkDirectiveRehype from "remark-directive-rehype";
import { realCtx } from "./mock";
import remarkDirectivePagelist from "../src/remarkDirectivePagelist";
import remarkDirectiveTiles from "../src/remarkDirectiveTiles";

export const toHtml = (md: string, ctx: HyperbookContext) => {
  const remarkPlugins: PluggableList = [
    remarkDirective,
    remarkDirectiveRehype,
    remarkDirectivePagelist(ctx),
    remarkDirectiveTiles(ctx),
  ];

  return unified()
    .use(remarkParse)
    .use(remarkPlugins)
    .use(remarkToRehype)
    .use(rehypeFormat)
    .use(rehypeStringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
    })
    .processSync(md);
};

describe("remarkDirectiveEmbed", () => {
  it("should transform", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="href(/elements/*)"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should sort by name asc", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="href(/elements/*)" orderBy="name:asc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should filter by name", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="name(.*IDE.*)" orderBy="name:asc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should sort by name desc", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="href(/elements/.*)" orderBy="name:desc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should filter glossary", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="href(/glossary/.*)" orderBy="name:desc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should filter glossary and format as glossary", async () => {
    expect(
      toHtml(
        `
::pagelist{format="glossary" source="href(/glossary/.*)"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should filter by keyword", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="keyword(test)" orderBy="name:desc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should filter by two keywords", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="keyword(test) AND keyword(ball)" orderBy="name:desc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should filter by one keyword and a href", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="keyword(test) AND href(/elements/.*)" orderBy="name:desc"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should format as ul", async () => {
    expect(
      toHtml(
        `
::pagelist{format="ul"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should format as ol", async () => {
    expect(
      toHtml(
        `
::pagelist{format="ol"}
`,
        realCtx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should register directives", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#list" source="href(/elements/*)"}
`,
        realCtx,
      ).data.directives?.["pagelist"],
    ).toBeDefined();
  });
  it("should transform directive in snippet", async () => {
    expect(
      toHtml(
        `
::pagelist{format="#portal" source="href(/elements/.*)" orderBy="index"}
`,
        realCtx,
      ),
    ).toMatchSnapshot();
  });
});
