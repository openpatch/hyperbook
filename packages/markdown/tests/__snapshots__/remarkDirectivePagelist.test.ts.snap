// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkDirectiveEmbed > should filter by keyword 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/elements/youtube">YouTube</a></li>
      <li><a href="/advanced/template-demo-yaml">Vorlagen - Demo</a></li>
      <li><a href="/advanced/templates">Vorlagen</a></li>
      <li><a href="/elements/video">Video</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should filter by name 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/elements/online-ide">Online IDE</a></li>
      <li><a href="/elements/sql-ide">SQL IDE</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should filter by one keyword and a href 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/elements/youtube">YouTube</a></li>
      <li><a href="/elements/video">Video</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should filter by two keywords 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/advanced/template-demo-yaml">Vorlagen - Demo</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should filter glossary 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/glossary/oop">OOP</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should filter glossary and format as glossary 1`] = `
"
<div class="directive-pagelist">
  <div class="container">
    <div class="letter">O</div>
    <ul class="terms">
      <li><a class="term" href="/glossary/oop">OOP</a></li>
    </ul>
  </div>
</div>
"
`;

exports[`remarkDirectiveEmbed > should format as ol 1`] = `
"
<div class="directive-pagelist">
  <ol>
    <li><a href="/elements/archive">Archiv</a></li>
    <li><a href="/elements/audio">Audio</a></li>
    <li><a href="/configuration/section">Bereich Konfiguration</a></li>
    <li><a href="/configuration/library">Bibliothek Konfiguration</a></li>
    <li><a href="/markdown/image">Bild</a></li>
    <li><a href="/elements/bitflow">Bitflow</a></li>
    <li><a href="/configuration/book">Buch Konfiguration</a></li>
    <li><a href="/elements/download">Download</a></li>
    <li><a href="/hosting/edugitpages">EduGit Pages</a></li>
    <li><a href="/hosting/custom">Eigene Lösung</a></li>
    <li><a href="/advanced/single-use-templates">Einweg Vorlagen</a></li>
    <li><a href="/elements">Elemente</a></li>
    <li><a href="/elements/embed">Embed</a></li>
    <li><a href="/elements/emoji">Emoji</a></li>
    <li><a href="/elements/excalidraw">Excalidraw</a></li>
    <li><a href="/advanced">Fortgeschrittene Features</a></li>
    <li><a href="/elements/protect">Geschützer Bereich</a></li>
    <li><a href="/hosting/ghpages">GitHub Pages</a></li>
    <li><a href="/hosting/glpages">GitLab Pages</a></li>
    <li><a href="/elements/glossary">Glossar</a></li>
    <li><a href="/markdown/emphasis">Hervorhebung</a></li>
    <li><a href="/elements/alert">Hinweise</a></li>
    <li><a href="/markdown/hr">Horizontale Linie</a></li>
    <li><a href="/hosting">Hosting</a></li>
    <li><a href="/elements/tiles">Kacheln</a></li>
    <li><a href="/elements/collapsible">Klappkasten</a></li>
    <li><a href="/markdown/comment">Kommentar</a></li>
    <li><a href="/configuration">Konfiguration</a></li>
    <li><a href="/elements/bookmarks">Lesezeichen</a></li>
    <li><a href="/markdown/link">Link</a></li>
    <li><a href="/markdown/list">Liste</a></li>
    <li><a href="/">Los Gehts</a></li>
    <li><a href="/markdown">Markdown Referenz</a></li>
    <li><a href="/elements/math">Mathmathik</a></li>
    <li><a href="/elements/mermaid">Mermaid</a></li>
    <li><a href="/elements/online-ide">Online IDE</a></li>
    <li><a href="/glossary/oop">OOP</a></li>
    <li><a href="/elements/plantuml">PlantUML</a></li>
    <li><a href="/elements/qr">QR Code</a></li>
    <li><a href="/markdown/code">Quelltext</a></li>
    <li><a href="/elements/tabs">Reiter</a></li>
    <li><a href="/elements/scratchblock">Scratchblöcke</a></li>
    <li><a href="/configuration/page">Seiten Konfiguration</a></li>
    <li><a href="/elements/slideshow">Slideshow</a></li>
    <li><a href="/elements/snippets">Snippets</a></li>
    <li><a href="/elements/sql-ide">SQL IDE</a></li>
    <li><a href="/elements/struktog">Struktog</a></li>
    <li><a href="/markdown/table">Tabelle</a></li>
    <li><a href="/markdown/header">Überschriften</a></li>
    <li><a href="/hosting/vercel">Vercel</a></li>
    <li><a href="/elements/video">Video</a></li>
    <li><a href="/advanced/template-demo-json">Vorlage - Demo</a></li>
    <li><a href="/advanced/templates">Vorlagen</a></li>
    <li><a href="/advanced/template-demo-yaml">Vorlagen - Demo</a></li>
    <li><a href="/elements/youtube">YouTube</a></li>
    <li><a href="/markdown/line-break">Zeilenumbruch</a></li>
    <li><a href="/markdown/quote">Zitat</a></li>
  </ol>
</div>
"
`;

exports[`remarkDirectiveEmbed > should format as ul 1`] = `
"
<div class="directive-pagelist">
  <ul>
    <li><a href="/elements/archive">Archiv</a></li>
    <li><a href="/elements/audio">Audio</a></li>
    <li><a href="/configuration/section">Bereich Konfiguration</a></li>
    <li><a href="/configuration/library">Bibliothek Konfiguration</a></li>
    <li><a href="/markdown/image">Bild</a></li>
    <li><a href="/elements/bitflow">Bitflow</a></li>
    <li><a href="/configuration/book">Buch Konfiguration</a></li>
    <li><a href="/elements/download">Download</a></li>
    <li><a href="/hosting/edugitpages">EduGit Pages</a></li>
    <li><a href="/hosting/custom">Eigene Lösung</a></li>
    <li><a href="/advanced/single-use-templates">Einweg Vorlagen</a></li>
    <li><a href="/elements">Elemente</a></li>
    <li><a href="/elements/embed">Embed</a></li>
    <li><a href="/elements/emoji">Emoji</a></li>
    <li><a href="/elements/excalidraw">Excalidraw</a></li>
    <li><a href="/advanced">Fortgeschrittene Features</a></li>
    <li><a href="/elements/protect">Geschützer Bereich</a></li>
    <li><a href="/hosting/ghpages">GitHub Pages</a></li>
    <li><a href="/hosting/glpages">GitLab Pages</a></li>
    <li><a href="/elements/glossary">Glossar</a></li>
    <li><a href="/markdown/emphasis">Hervorhebung</a></li>
    <li><a href="/elements/alert">Hinweise</a></li>
    <li><a href="/markdown/hr">Horizontale Linie</a></li>
    <li><a href="/hosting">Hosting</a></li>
    <li><a href="/elements/tiles">Kacheln</a></li>
    <li><a href="/elements/collapsible">Klappkasten</a></li>
    <li><a href="/markdown/comment">Kommentar</a></li>
    <li><a href="/configuration">Konfiguration</a></li>
    <li><a href="/elements/bookmarks">Lesezeichen</a></li>
    <li><a href="/markdown/link">Link</a></li>
    <li><a href="/markdown/list">Liste</a></li>
    <li><a href="/">Los Gehts</a></li>
    <li><a href="/markdown">Markdown Referenz</a></li>
    <li><a href="/elements/math">Mathmathik</a></li>
    <li><a href="/elements/mermaid">Mermaid</a></li>
    <li><a href="/elements/online-ide">Online IDE</a></li>
    <li><a href="/glossary/oop">OOP</a></li>
    <li><a href="/elements/plantuml">PlantUML</a></li>
    <li><a href="/elements/qr">QR Code</a></li>
    <li><a href="/markdown/code">Quelltext</a></li>
    <li><a href="/elements/tabs">Reiter</a></li>
    <li><a href="/elements/scratchblock">Scratchblöcke</a></li>
    <li><a href="/configuration/page">Seiten Konfiguration</a></li>
    <li><a href="/elements/slideshow">Slideshow</a></li>
    <li><a href="/elements/snippets">Snippets</a></li>
    <li><a href="/elements/sql-ide">SQL IDE</a></li>
    <li><a href="/elements/struktog">Struktog</a></li>
    <li><a href="/markdown/table">Tabelle</a></li>
    <li><a href="/markdown/header">Überschriften</a></li>
    <li><a href="/hosting/vercel">Vercel</a></li>
    <li><a href="/elements/video">Video</a></li>
    <li><a href="/advanced/template-demo-json">Vorlage - Demo</a></li>
    <li><a href="/advanced/templates">Vorlagen</a></li>
    <li><a href="/advanced/template-demo-yaml">Vorlagen - Demo</a></li>
    <li><a href="/elements/youtube">YouTube</a></li>
    <li><a href="/markdown/line-break">Zeilenumbruch</a></li>
    <li><a href="/markdown/quote">Zitat</a></li>
  </ul>
</div>
"
`;

exports[`remarkDirectiveEmbed > should sort by name asc 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/elements/archive">Archiv</a></li>
      <li><a href="/elements/audio">Audio</a></li>
      <li><a href="/elements/bitflow">Bitflow</a></li>
      <li><a href="/elements/download">Download</a></li>
      <li><a href="/elements">Elemente</a></li>
      <li><a href="/elements/embed">Embed</a></li>
      <li><a href="/elements/emoji">Emoji</a></li>
      <li><a href="/elements/excalidraw">Excalidraw</a></li>
      <li><a href="/elements/protect">Geschützer Bereich</a></li>
      <li><a href="/elements/glossary">Glossar</a></li>
      <li><a href="/elements/alert">Hinweise</a></li>
      <li><a href="/elements/tiles">Kacheln</a></li>
      <li><a href="/elements/collapsible">Klappkasten</a></li>
      <li><a href="/elements/bookmarks">Lesezeichen</a></li>
      <li><a href="/elements/math">Mathmathik</a></li>
      <li><a href="/elements/mermaid">Mermaid</a></li>
      <li><a href="/elements/online-ide">Online IDE</a></li>
      <li><a href="/elements/plantuml">PlantUML</a></li>
      <li><a href="/elements/qr">QR Code</a></li>
      <li><a href="/elements/tabs">Reiter</a></li>
      <li><a href="/elements/scratchblock">Scratchblöcke</a></li>
      <li><a href="/elements/slideshow">Slideshow</a></li>
      <li><a href="/elements/snippets">Snippets</a></li>
      <li><a href="/elements/sql-ide">SQL IDE</a></li>
      <li><a href="/elements/struktog">Struktog</a></li>
      <li><a href="/elements/video">Video</a></li>
      <li><a href="/elements/youtube">YouTube</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should sort by name desc 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/elements/youtube">YouTube</a></li>
      <li><a href="/elements/video">Video</a></li>
      <li><a href="/elements/struktog">Struktog</a></li>
      <li><a href="/elements/sql-ide">SQL IDE</a></li>
      <li><a href="/elements/snippets">Snippets</a></li>
      <li><a href="/elements/slideshow">Slideshow</a></li>
      <li><a href="/elements/scratchblock">Scratchblöcke</a></li>
      <li><a href="/elements/tabs">Reiter</a></li>
      <li><a href="/elements/qr">QR Code</a></li>
      <li><a href="/elements/plantuml">PlantUML</a></li>
      <li><a href="/elements/online-ide">Online IDE</a></li>
      <li><a href="/elements/mermaid">Mermaid</a></li>
      <li><a href="/elements/math">Mathmathik</a></li>
      <li><a href="/elements/bookmarks">Lesezeichen</a></li>
      <li><a href="/elements/collapsible">Klappkasten</a></li>
      <li><a href="/elements/tiles">Kacheln</a></li>
      <li><a href="/elements/alert">Hinweise</a></li>
      <li><a href="/elements/glossary">Glossar</a></li>
      <li><a href="/elements/protect">Geschützer Bereich</a></li>
      <li><a href="/elements/excalidraw">Excalidraw</a></li>
      <li><a href="/elements/emoji">Emoji</a></li>
      <li><a href="/elements/embed">Embed</a></li>
      <li><a href="/elements/download">Download</a></li>
      <li><a href="/elements/bitflow">Bitflow</a></li>
      <li><a href="/elements/audio">Audio</a></li>
      <li><a href="/elements/archive">Archiv</a></li>
    </ul>
  
</div>
"
`;

exports[`remarkDirectiveEmbed > should transform 1`] = `
"
<div class="directive-pagelist">
  
    <ul>
      <li><a href="/elements/archive">Archiv</a></li>
      <li><a href="/elements/audio">Audio</a></li>
      <li><a href="/elements/bitflow">Bitflow</a></li>
      <li><a href="/elements/download">Download</a></li>
      <li><a href="/elements">Elemente</a></li>
      <li><a href="/elements/embed">Embed</a></li>
      <li><a href="/elements/emoji">Emoji</a></li>
      <li><a href="/elements/excalidraw">Excalidraw</a></li>
      <li><a href="/elements/protect">Geschützer Bereich</a></li>
      <li><a href="/elements/glossary">Glossar</a></li>
      <li><a href="/elements/alert">Hinweise</a></li>
      <li><a href="/elements/tiles">Kacheln</a></li>
      <li><a href="/elements/collapsible">Klappkasten</a></li>
      <li><a href="/elements/bookmarks">Lesezeichen</a></li>
      <li><a href="/elements/math">Mathmathik</a></li>
      <li><a href="/elements/mermaid">Mermaid</a></li>
      <li><a href="/elements/online-ide">Online IDE</a></li>
      <li><a href="/elements/plantuml">PlantUML</a></li>
      <li><a href="/elements/qr">QR Code</a></li>
      <li><a href="/elements/tabs">Reiter</a></li>
      <li><a href="/elements/scratchblock">Scratchblöcke</a></li>
      <li><a href="/elements/slideshow">Slideshow</a></li>
      <li><a href="/elements/snippets">Snippets</a></li>
      <li><a href="/elements/sql-ide">SQL IDE</a></li>
      <li><a href="/elements/struktog">Struktog</a></li>
      <li><a href="/elements/video">Video</a></li>
      <li><a href="/elements/youtube">YouTube</a></li>
    </ul>
  
</div>
"
`;
